File: app/src/main/java/com/example/engine/SimplexNoise.kt
package com.example.engine

import kotlin.math.floor

/**
 * Fast Simplex & Fractal Perlin-style Noise Generator
 */
class SimplexNoise(seed: Long = 1337L) {

    private val p = IntArray(512)
    private val perm = IntArray(512)
    private val permMod12 = IntArray(512)

    init {
        val rand = java.util.Random(seed)
        val source = (0..255).toList().shuffled(rand)
        for (i in 0..255) {
            p[i] = source[i]
            p[256 + i] = source[i]
            perm[i] = source[i]
            perm[256 + i] = source[i]
            permMod12[i] = (source[i] % 12)
            permMod12[256 + i] = (source[i] % 12)
        }
    }

    private val grad3 = arrayOf(
        floatArrayOf(1f, 1f, 0f), floatArrayOf(-1f, 1f, 0f), floatArrayOf(1f, -1f, 0f), floatArrayOf(-1f, -1f, 0f),
        floatArrayOf(1f, 0f, 1f), floatArrayOf(-1f, 0f, 1f), floatArrayOf(1f, 0f, -1f), floatArrayOf(-1f, 0f, -1f),
        floatArrayOf(0f, 1f, 1f), floatArrayOf(0f, -1f, 1f), floatArrayOf(0f, 1f, -1f), floatArrayOf(0f, -1f, -1f)
    )

    private fun dot(g: FloatArray, x: Float, y: Float): Float = g[0] * x + g[1] * y
    private fun dot(g: FloatArray, x: Float, y: Float, z: Float): Float = g[0] * x + g[1] * y + g[2] * z

    fun eval2D(xin: Float, yin: Float): Float {
        val f2 = 0.5f * (Math.sqrt(3.0) - 1.0).toFloat()
        val g2 = (3.0 - Math.sqrt(3.0)).toFloat() / 6.0f

        val s = (xin + yin) * f2
        val i = floor(xin + s).toInt()
        val j = floor(yin + s).toInt()
        val t = (i + j) * g2
        val x0 = xin - (i - t)
        val y0 = yin - (j - t)

        val i1: Int
        val j1: Int
        if (x0 > y0) {
            i1 = 1; j1 = 0
        } else {
            i1 = 0; j1 = 1
        }

        val x1 = x0 - i1 + g2
        val y1 = y0 - j1 + g2
        val x2 = x0 - 1.0f + 2.0f * g2
        val y2 = y0 - 1.0f + 2.0f * g2

        val ii = i and 255
        val jj = j and 255
        val gi0 = permMod12[ii + perm[jj]]
        val gi1 = permMod12[ii + i1 + perm[jj + j1]]
        val gi2 = permMod12[ii + 1 + perm[jj + 1]]

        var n0 = 0f
        var n1 = 0f
        var n2 = 0f

        var t0 = 0.5f - x0 * x0 - y0 * y0
        if (t0 > 0) {
            t0 *= t0
            n0 = t0 * t0 * dot(grad3[gi0], x0, y0)
        }

        var t1 = 0.5f - x1 * x1 - y1 * y1
        if (t1 > 0) {
            t1 *= t1
            n1 = t1 * t1 * dot(grad3[gi1], x1, y1)
        }

        var t2 = 0.5f - x2 * x2 - y2 * y2
        if (t2 > 0) {
            t2 *= t2
            n2 = t2 * t2 * dot(grad3[gi2], x2, y2)
        }

        return 70.0f * (n0 + n1 + n2)
    }

    fun eval3D(xin: Float, yin: Float, zin: Float): Float {
        val f3 = 1.0f / 3.0f
        val s = (xin + yin + zin) * f3
        val i = floor(xin + s).toInt()
        val j = floor(yin + s).toInt()
        val k = floor(zin + s).toInt()

        val g3 = 1.0f / 6.0f
        val t = (i + j + k) * g3
        val x0 = xin - (i - t)
        val y0 = yin - (j - t)
        val z0 = zin - (k - t)

        val i1: Int; val j1: Int; val k1: Int
        val i2: Int; val j2: Int; val k2: Int
        if (x0 >= y0) {
            if (y0 >= z0) {
                i1 = 1; j1 = 0; k1 = 0; i2 = 1; j2 = 1; k2 = 0
            } else if (x0 >= z0) {
                i1 = 1; j1 = 0; k1 = 0; i2 = 1; j2 = 0; k2 = 1
            } else {
                i1 = 0; j1 = 0; k1 = 1; i2 = 1; j2 = 0; k2 = 1
            }
        } else {
            if (y0 < z0) {
                i1 = 0; j1 = 0; k1 = 1; i2 = 0; j2 = 1; k2 = 1
            } else if (x0 < z0) {
                i1 = 0; j1 = 1; k1 = 0; i2 = 0; j2 = 1; k2 = 1
            } else {
                i1 = 0; j1 = 1; k1 = 0; i2 = 1; j2 = 1; k2 = 0
            }
        }

        val x1 = x0 - i1 + g3
        val y1 = y0 - j1 + g3
        val z1 = z0 - k1 + g3
        val x2 = x0 - i2 + 2.0f * g3
        val y2 = y0 - j2 + 2.0f * g3
        val z2 = z0 - k2 + 2.0f * g3
        val x3 = x0 - 1.0f + 3.0f * g3
        val y3 = y0 - 1.0f + 3.0f * g3
        val z3 = z0 - 1.0f + 3.0f * g3

        val ii = i and 255
        val jj = j and 255
        val kk = k and 255

        val gi0 = permMod12[ii + perm[jj + perm[kk]]]
        val gi1 = permMod12[ii + i1 + perm[jj + j1 + perm[kk + k1]]]
        val gi2 = permMod12[ii + i2 + perm[jj + j2 + perm[kk + k2]]]
        val gi3 = permMod12[ii + 1 + perm[jj + 1 + perm[kk + 1]]]

        var n0 = 0f; var n1 = 0f; var n2 = 0f; var n3 = 0f

        var t0 = 0.6f - x0 * x0 - y0 * y0 - z0 * z0
        if (t0 > 0) {
            t0 *= t0
            n0 = t0 * t0 * dot(grad3[gi0], x0, y0, z0)
        }

        var t1 = 0.6f - x1 * x1 - y1 * y1 - z1 * z1
        if (t1 > 0) {
            t1 *= t1
            n1 = t1 * t1 * dot(grad3[gi1], x1, y1, z1)
        }

        var t2 = 0.6f - x2 * x2 - y2 * y2 - z2 * z2
        if (t2 > 0) {
            t2 *= t2
            n2 = t2 * t2 * dot(grad3[gi2], x2, y2, z2)
        }

        var t3 = 0.6f - x3 * x3 - y3 * y3 - z3 * z3
        if (t3 > 0) {
            t3 *= t3
            n3 = t3 * t3 * dot(grad3[gi3], x3, y3, z3)
        }

        return 32.0f * (n0 + n1 + n2 + n3)
    }

    /**
     * Fractal Octave Noise
     */
    fun fractal2D(x: Float, y: Float, octaves: Int = 4, persistence: Float = 0.5f, lacunarity: Float = 2.0f): Float {
        var total = 0f
        var frequency = 1.0f
        var amplitude = 1.0f
        var maxValue = 0f
        for (i in 0 until octaves) {
            total += eval2D(x * frequency, y * frequency) * amplitude
            maxValue += amplitude
            amplitude *= persistence
            frequency *= lacunarity
        }
        return total / maxValue
    }
}

File: js/simplex-noise.js
/**
 * Fast Simplex & Fractal Perlin-style Noise Generator
 * Converted from SimplexNoise.kt
 */
class SimplexNoise {
  constructor(seed = 1337) {
    this.p = new Int32Array(512);
    this.perm = new Int32Array(512);
    this.permMod12 = new Int32Array(512);

    // Simple pseudo-random seeded permutation generator
    let source = Array.from({ length: 256 }, (_, i) => i);
    let s = seed % 2147483647;
    if (s <= 0) s += 2147483646;

    const nextRand = () => {
      s = (s * 16807) % 2147483647;
      return (s - 1) / 2147483646;
    };

    // Shuffle source array
    for (let i = 255; i > 0; i--) {
      const j = Math.floor(nextRand() * (i + 1));
      [source[i], source[j]] = [source[j], source[i]];
    }

    for (let i = 0; i < 256; i++) {
      this.p[i] = source[i];
      this.p[256 + i] = source[i];
      this.perm[i] = source[i];
      this.perm[256 + i] = source[i];
      this.permMod12[i] = source[i] % 12;
      this.permMod12[256 + i] = source[i] % 12;
    }

    this.grad3 = [
      [1, 1, 0], [-1, 1, 0], [1, -1, 0], [-1, -1, 0],
      [1, 0, 1], [-1, 0, 1], [1, 0, -1], [-1, 0, -1],
      [0, 1, 1], [0, -1, 1], [0, 1, -1], [0, -1, -1]
    ];
  }

  dot2D(g, x, y) {
    return g[0] * x + g[1] * y;
  }

  dot3D(g, x, y, z) {
    return g[0] * x + g[1] * y + g[2] * z;
  }

  eval2D(xin, yin) {
    const f2 = 0.5 * (Math.sqrt(3.0) - 1.0);
    const g2 = (3.0 - Math.sqrt(3.0)) / 6.0;

    const s = (xin + yin) * f2;
    const i = Math.floor(xin + s);
    const j = Math.floor(yin + s);
    const t = (i + j) * g2;
    const x0 = xin - (i - t);
    const y0 = yin - (j - t);

    let i1, j1;
    if (x0 > y0) {
      i1 = 1; j1 = 0;
    } else {
      i1 = 0; j1 = 1;
    }

    const x1 = x0 - i1 + g2;
    const y1 = y0 - j1 + g2;
    const x2 = x0 - 1.0 + 2.0 * g2;
    const y2 = y0 - 1.0 + 2.0 * g2;

    const ii = i & 255;
    const jj = j & 255;
    const gi0 = this.permMod12[ii + this.perm[jj]];
    const gi1 = this.permMod12[ii + i1 + this.perm[jj + j1]];
    const gi2 = this.permMod12[ii + 1 + this.perm[jj + 1]];

    let n0 = 0, n1 = 0, n2 = 0;

    let t0 = 0.5 - x0 * x0 - y0 * y0;
    if (t0 > 0) {
      t0 *= t0;
      n0 = t0 * t0 * this.dot2D(this.grad3[gi0], x0, y0);
    }

    let t1 = 0.5 - x1 * x1 - y1 * y1;
    if (t1 > 0) {
      t1 *= t1;
      n1 = t1 * t1 * this.dot2D(this.grad3[gi1], x1, y1);
    }

    let t2 = 0.5 - x2 * x2 - y2 * y2;
    if (t2 > 0) {
      t2 *= t2;
      n2 = t2 * t2 * this.dot2D(this.grad3[gi2], x2, y2);
    }

    return 70.0 * (n0 + n1 + n2);
  }

  eval3D(xin, yin, zin) {
    const f3 = 1.0 / 3.0;
    const s = (xin + yin + zin) * f3;
    const i = Math.floor(xin + s);
    const j = Math.floor(yin + s);
    const k = Math.floor(zin + s);

    const g3 = 1.0 / 6.0;
    const t = (i + j + k) * g3;
    const x0 = xin - (i - t);
    const y0 = yin - (j - t);
    const z0 = zin - (k - t);

    let i1, j1, k1;
    let i2, j2, k2;
    if (x0 >= y0) {
      if (y0 >= z0) {
        i1 = 1; j1 = 0; k1 = 0; i2 = 1; j2 = 1; k2 = 0;
      } else if (x0 >= z0) {
        i1 = 1; j1 = 0; k1 = 0; i2 = 1; j2 = 0; k2 = 1;
      } else {
        i1 = 0; j1 = 0; k1 = 1; i2 = 1; j2 = 0; k2 = 1;
      }
    } else {
      if (y0 < z0) {
        i1 = 0; j1 = 0; k1 = 1; i2 = 0; j2 = 1; k2 = 1;
      } else if (x0 < z0) {
        i1 = 0; j1 = 1; k1 = 0; i2 = 0; j2 = 1; k2 = 1;
      } else {
        i1 = 0; j1 = 1; k1 = 0; i2 = 1; j2 = 1; k2 = 0;
      }
    }

    const x1 = x0 - i1 + g3;
    const y1 = y0 - j1 + g3;
    const z1 = z0 - k1 + g3;
    const x2 = x0 - i2 + 2.0 * g3;
    const y2 = y0 - j2 + 2.0 * g3;
    const z2 = z0 - k2 + 2.0 * g3;
    const x3 = x0 - 1.0 + 3.0 * g3;
    const y3 = y0 - 1.0 + 3.0 * g3;
    const z3 = z0 - 1.0 + 3.0 * g3;

    const ii = i & 255;
    const jj = j & 255;
    const kk = k & 255;

    const gi0 = this.permMod12[ii + this.perm[jj + this.perm[kk]]];
    const gi1 = this.permMod12[ii + i1 + this.perm[jj + j1 + this.perm[kk + k1]]];
    const gi2 = this.permMod12[ii + i2 + this.perm[jj + j2 + this.perm[kk + k2]]];
    const gi3 = this.permMod12[ii + 1 + this.perm[jj + 1 + this.perm[kk + 1]]];

    let n0 = 0, n1 = 0, n2 = 0, n3 = 0;

    let t0 = 0.6 - x0 * x0 - y0 * y0 - z0 * z0;
    if (t0 > 0) {
      t0 *= t0;
      n0 = t0 * t0 * this.dot3D(this.grad3[gi0], x0, y0, z0);
    }

    let t1 = 0.6 - x1 * x1 - y1 * y1 - z1 * z1;
    if (t1 > 0) {
      t1 *= t1;
      n1 = t1 * t1 * this.dot3D(this.grad3[gi1], x1, y1, z1);
    }

    let t2 = 0.6 - x2 * x2 - y2 * y2 - z2 * z2;
    if (t2 > 0) {
      t2 *= t2;
      n2 = t2 * t2 * this.dot3D(this.grad3[gi2], x2, y2, z2);
    }

    let t3 = 0.6 - x3 * x3 - y3 * y3 - z3 * z3;
    if (t3 > 0) {
      t3 *= t3;
      n3 = t3 * t3 * this.dot3D(this.grad3[gi3], x3, y3, z3);
    }

    return 32.0 * (n0 + n1 + n2 + n3);
  }

  fractal2D(x, y, octaves = 4, persistence = 0.5, lacunarity = 2.0) {
    let total = 0;
    let frequency = 1.0;
    let amplitude = 1.0;
    let maxValue = 0;
    for (let i = 0; i < octaves; i++) {
      total += this.eval2D(x * frequency, y * frequency) * amplitude;
      maxValue += amplitude;
      amplitude *= persistence;
      frequency *= lacunarity;
    }
    return total / maxValue;
  }
}
